File: install/TileAtlasCreator/src/main/java/mdcii/PaletteFile.java
// This file is part of the MDCII project.
//
// Copyright (c) 2023. stwe <https://github.com/stwe/MDCII>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

package mdcii;

import java.io.IOException;
import java.nio.file.Path;
import java.util.Objects;

import static mdcii.Log.LOGGER;

/**
 * Represents a PaletteFile.
 * The class loads the pallet values from the <i>stadtfld.col</i> file.
 */
public class PaletteFile extends BinaryFile {

    //-------------------------------------------------
    // Constants
    //-------------------------------------------------

    /**
     * The number of palette colors.
     */
    private static final int NUMBER_OF_COLORS = 256;

    /**
     * The number of expected {@link Chunk} objects in the file.
     */
    private static final int NUMBER_OF_CHUNKS = 1;

    /**
     * The ID of the {@link Chunk}.
     */
    private static final String CHUNK_ID = "COL";

    //-------------------------------------------------
    // Member
    //-------------------------------------------------

    /**
     * An array with the palette color values.
     * The color values are saved as int.
     */
    private final int[] palette = new int[NUMBER_OF_COLORS];

    //-------------------------------------------------
    // Ctors.
    //-------------------------------------------------

    /**
     * Constructs a new {@link PaletteFile} object.
     *
     * @param path The {@link Path} to the palette file.
     * @throws IOException If an I/O error is thrown.
     */
    PaletteFile(Path path) throws IOException {
        super(Objects.requireNonNull(path, "path must not be null"));

        LOGGER.debug("Creates PaletteFile object from file {}.", path);

        if (getNumberOfChunks() != NUMBER_OF_CHUNKS) {
            throw new RuntimeException("Invalid number of Chunks.");
        }

        if (!chunkIndexHasId(0, CHUNK_ID)) {
            throw new RuntimeException("Invalid Chunk Id.");
        }

        readDataFromChunks();
    }

    //-------------------------------------------------
    // Getter
    //-------------------------------------------------

    /**
     * Get {@link #palette}.
     *
     * @return {@link #palette}
     */
    public int[] getPalette() {
        return palette;
    }

    //-------------------------------------------------
    // BinaryFileInterface
    //-------------------------------------------------

    @Override
    public void readDataFromChunks() {
        LOGGER.debug("Start reading Palette data from Chunks...");

        var chunk0 = getChunk(0);

        for (int i = 0; i < palette.length; i++) {
            int red = Util.byteToInt(chunk0.getData().get());
            int green = Util.byteToInt(chunk0.getData().get());
            int blue = Util.byteToInt(chunk0.getData().get());

            chunk0.getData().get(); // skip next byte

            palette[i] = Util.rgbToInt(red, green, blue);
        }

        LOGGER.debug("Palette data read successfully.");
    }
}

File: src/resource/PaletteFile.cpp
// This file is part of the MDCII project.
//
// Copyright (c) 2023. stwe <https://github.com/stwe/MDCII>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

#include "PaletteFile.h"
#include "MdciiAssert.h"
#include "MdciiException.h"
#include "chunk/Chunk.h"

//-------------------------------------------------
// Ctors. / Dtor.
//-------------------------------------------------

mdcii::resource::PaletteFile::PaletteFile(std::string t_filePath)
    : BinaryFile(std::move(t_filePath))
{
    MDCII_LOG_DEBUG("[PaletteFile::PaletteFile()] Create PaletteFile.");

    if (chunks.at(0)->id.c_str() != CHUNK_ID) // NOLINT(*-redundant-string-cstr)
    {
        throw MDCII_EXCEPTION("[PaletteFile::PaletteFile()] Invalid Chunk Id.");
    }
}

mdcii::resource::PaletteFile::~PaletteFile() noexcept
{
    MDCII_LOG_DEBUG("[PaletteFile::~PaletteFile()] Destruct PaletteFile.");
}

//-------------------------------------------------
// BinaryFile Interface
//-------------------------------------------------

void mdcii::resource::PaletteFile::ReadDataFromChunks()
{
    MDCII_LOG_DEBUG("[PaletteFile::ReadDataFromChunks()] Start reading Palette data from Chunks...");

    const auto& chunk0{ chunks.at(0) };

    for (size_t i{ 0 }; i < chunk0->length; i += 4)
    {
        const auto red{ chunk0->data.at(i) };
        const auto green{ chunk0->data.at(i + 1) };
        const auto blue{ chunk0->data.at(i + 2) };

        palette.emplace_back(red, green, blue);
    }

    MDCII_ASSERT(palette.size() == NUMBER_OF_COLORS, "[PaletteFile::ReadDataFromChunks()] Invalid number of colors.")

    MDCII_LOG_DEBUG("[PaletteFile::ReadDataFromChunks()] Palette data read successfully.");
}
